File: AlgoLeaf/2dGeneration/math_stuff.py
from scipy.spatial import distance


def translation(point, direction, stepsize):
    """

    :param point: coord of the particle we want to translate
    :param dir_petiole: Unit vector to the petiole
    :param dir_neighbour: Unit vector to the neighbour
    :param s: let us control the translation
    :return: The coord of the point translated by the two vectors
    """
    x = point[0] + stepsize * direction[0]
    y = point[1] + stepsize * direction[1]
    return [x, y]


def compute_dir(dir_t, dir_n, weight_n, weight_t):
    """

    :param dir_t: vector direction to target
    :param dir_n: vector direction to neighbour
    :param weight_n: weight of neighbout
    :param weight_t: weight of target
    :return: normalise vector to the direction
    """

    x = (dir_t[0] * weight_t + dir_n[0] * weight_n) / (weight_t + weight_n)
    y = (dir_t[1] * weight_t + dir_n[1] * weight_n) / (weight_t + weight_n)
    norme = (x**2 + y**2)**0.5
    return [x / norme, y /norme]


def get_unit_vector(point1, point2):
    """

    :param point1: first point
    :param point2: second point
    :return: A unit vector computed with the two points
    """
    x = point2[0] - point1[0]
    y = point2[1] - point1[1]
    norme = (x**2 + y**2)**0.5
    return [x / norme, y / norme]


def nearest_neigbour(particle, particle_set):
    """

    :param particle: coord of the particle we want to find the neighbour
    :param particle_set: list of all the coord of all particles
    :return: the coordinate of the nearest particle of "particle"
    """
    closest_index = distance.cdist([particle.coord], [x.coord for x in particle_set]).argmin()

    return particle_set[closest_index]


def get_distance(particule1, particule2):
    """

    :param particule1: a particle
    :param particule2: another particle
    :return: distance between both particle
    """
    x = particule2.coord[0] - particule1.coord[0]
    y = particule2.coord[1] - particule1.coord[1]

    return (x**2 + y**2)**0.5


def merge_pos(particule1, particule2):
    """

    :param particule1: a particle
    :param particule2: another particle
    :return: the mean distance between both
    """

    x = (particule1.coord[0] + particule2.coord[0]) / 2
    y = (particule1.coord[1] + particule2.coord[1]) / 2

    return [x, y]

File: Cpp/src/AlgoLeaf/math_stuff.cpp
//
// Created by dany on 24/03/19.
//

#include <cmath>
#include "math_stuff.hh"


namespace math {

    Point3D math::get_translation(Point3D p, Vector3D dir, float stepsize){


        double x = p.getX() + stepsize * dir.getX();
        double y = p.getY() + stepsize * dir.getY();
        double z = p.getZ() + stepsize * dir.getZ();

        return Point3D(x, y , z);


    }


    Vector3D math::compute_dir(Vector3D dir_t, Vector3D dir_n, float weight_n, float weight_t){

        double x = (dir_t.getX() * weight_t + dir_n.getX() * weight_n) / (weight_t + weight_n);
        double y = (dir_t.getY() * weight_t + dir_n.getY() * weight_n) / (weight_t + weight_n);
        double z = (dir_t.getY() * weight_t + dir_n.getY() * weight_n) / (weight_t + weight_n);
        double norme = sqrt(x*x + y*y + z*z);

        return Vector3D(x / norme, y / norme, z / norme);


    }


    Vector3D math::get_unit_vector(Point3D p1, Point3D p2){

        double x = p2.getX() - p1.getX();
        double y = p2.getY() - p1.getY();
        double z = p2.getZ() - p1.getZ();

        double norme = sqrt(x*x + y*y + z*z);

        return Vector3D(x / norme, y / norme, z / norme);


    }

    /*
    particles::Particle math::get_neighbour(particles::Particle p, particles::Particle_set set){



    }
    */

    double math::get_distance(particles::Particle p1, particles::Particle p2) {

        double x = p2.position.getX() - p1.position.getX();
        double y = p2.position.getY() - p1.position.getY();
        double z = p2.position.getZ() - p1.position.getZ();

        return sqrt(x*x + y*y + z*z);
    }

    Point3D math::merge_pos(particles::Particle p1, particles::Particle p2){

        double x = (p1.position.getX() + p2.position.getX()) / 2;
        double y = (p1.position.getY() + p2.position.getY()) / 2;
        double z = (p1.position.getZ() + p2.position.getZ()) / 2;

        return Point3D(x, y , z);

    }




}

